pkg/pillar/cmd/domainmgr/containerd.go

```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
+	}
+
+	_, err = ctrCreateNewTask(containerID)
+	if err != nil {
+		log.Errorf("Could not create a new task for container: %v. %v", containerID, err.Error())
+		return "", fmt.Errorf("Could not create a new task for container: %v. %v", containerID, err.Error())
+	}
+	return containerID, nil
+}
+
+// ctrCreateNewTask - Creates a new task for the container.
+// This step is responsible for setting up rootfs of the container under /run/containerd/<RunTime>/<NameSpace>/<Container id>/
+// NOTE: As per containerd's flow, to start a container, first a task is created which sets up container's rootfs,
+// then the task is started which invokes runc binary with the rootfs. Since we are not using runc, we need to stop at task create step.
+func ctrCreateNewTask(containerID string) (containerd.Task, error) {
+	var err error
+	ctrdClient, err := getContainerdClient(ctrdSocket)
+	if err != nil {
+		log.Errorf("Could not build containerd client. %v", err.Error())
+		return nil, fmt.Errorf("Exception while loading container. %v", err)
+	}
+	defer ctrdClient.Close()
+	ctrdCtx := getContainerdContext(CtrdServicesNamespace)
+
+	container, err := loadContainer(containerID)
+	if err != nil {
+		return nil, err
+	}
+
+	return container.NewTask(ctrdCtx, cio.NewCreator(cio.WithStdio))
 }
 
 // ctrPrepare prepare an existing container
 func ctrPrepare(containerImageSha256 string, status *types.DomainStatus) (string, error) {
-	ctrID := uuid.NewV4().String()
+	containerID := buildContainerID(status.DisplayName)
 	ociFilename, err := utils.VerifiedImageFileLocation(containerImageSha256)
 	if err != nil {
 		log.Errorf("failed to get Image File Location. "+
 			"err: %+s", err.Error())
-		return ctrID, err
+		return containerID, err
 	}
 	log.Infof("ociFilename %s sha %s", ociFilename, containerImageSha256)
 
 	loadedImages, err := containerdLoadImageTar(ociFilename)
 	if err != nil {
 		log.Errorf("failed to load Image File at %s into containerd: %+s", ociFilename, err.Error())
-		return ctrID, err
+		return containerID, err
 	}
 
 	// we currently only support one image per file; will change eventually
 	if len(loadedImages) != 1 {
 		log.Errorf("loaded %d images, expected just 1", len(loadedImages))
 	}
-	var imageHash string
-	for _, hash := range loadedImages {
-		imageHash = hash
+	var imageName string
+	for _, name := range loadedImages {
+		imageName = name
 	}
 
-	// TODO:
-	// 1. create a container using containerd - remember that containerd does not
-	//    create a name for it, so we need to do so. Docker uses a hash of the config.
-	//    We can just create a UUID.
-	// 2. create a task using containerd - this will create the snapshot
-	// 3. include any necessary mountpoints
-	if err = ctrCreate(ctrID, imageHash); err != nil {
-		return ctrID, fmt.Errorf("failed to create container %s: %v", ctrID, imageHash)
+	if containerID, err := ctrCreate(containerID, imageName, status); err != nil {
```

Comment:
This is strange and non-idiomatic. `containerD := ctrCreate(containerID,...)`? We should not be overriding the var, especially as it is the same value. Linter should catch this. Just make it `if err = ctrCreate(...); err != nil {`

Reply:
Corrected.